### PR TITLE
Vpd tool read keyword

### PR DESCRIPTION
**Pull request by branupama**

### CODE REVIEW

File: vpd-tool/include/utils.hpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include <iostream>
+
+namespace vpd
+{
+namespace utils
+{
+/**
+ * @brief An API to read property from Dbus.
+ *
+ * The caller of the API needs to validate the validatity and correctness of the
```
Comment (souvik1914581) [inline vpd-tool/include/utils.hpp:18]: API description can be more descriptive. Some typos in this text.

File: vpd-tool/include/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * @brief An API to print json data on stdout.
+ *
+ * @param[in] i_jsonData - JSON object.
+ */
+inline void printJson(const nlohmann::json& i_jsonData)
+{
+    try
+    {
+        std::cout << i_jsonData.dump(4) << std::endl;
```
Comment (souvik1914581) [inline vpd-tool/include/utils.hpp:78]: nit: Can use a `constexpr` variable to name the indentation value `4` passed to `dump()` for improved readability.

File: vpd-tool/include/utils.hpp (hunk 1 above)
Comment (branupama) [reply vpd-tool/include/utils.hpp:18]: done

File: include/utility/dbus_utility.hpp [hunk 3]
```diff
@@ -106,8 +106,8 @@ inline types::PropertyMap getPropertyMap(const std::string& i_service,
 /**
  * @brief An API to read property from Dbus.
  *
- * The caller of the API needs to validate the validatity and correctness of the
- * type and value of data returned. The API will just fetch and retun the data
+ * The caller of the API needs to validate the validity and correctness of the
```
Comment (PriyangaRamasamy) [inline include/utility/dbus_utility.hpp:109]: why spell correction change got included in this PR? please revert it. Be specific to vpd-tool read implementation only.

File: vpd-tool/include/utils.hpp (hunk 1 above)
Comment (PriyangaRamasamy) [reply vpd-tool/include/utils.hpp:21]: validate the validity and correctness - all the 3 words points to same meaning.

File: vpd-tool/src/vpd_tool.cpp [hunk 4]
```diff
@@ -8,13 +14,60 @@ int VpdTool::readKeyword(const std::string& i_fruPath,
...
+            l_keywordValue = utils::readDbusProperty(
+                constants::inventoryManagerService, l_inventoryObjectPath,
+                constants::ipzVpdInf + i_recordName, i_keywordName);
+        }
+
+        if (const auto l_value =
+                std::get_if<types::BinaryVector>(&l_keywordValue))
+        {
+            const std::string& l_keywordStrValue =
+                utils::getPrintableValue(*l_value);
```
Comment (PriyangaRamasamy) [inline vpd-tool/src/vpd_tool.cpp:39]: As part of code refactoring, we have decided to print value in both the formats - HEX and ASCII. Please revisit on this

File: vpd-tool/src/vpd_tool.cpp [hunk 5]
```diff
@@ -1,5 +1,11 @@
 #include "vpd_tool.hpp"
 
+#include "tool_constants.hpp"
+#include "tool_types.hpp"
+#include "utils.hpp"
```
Comment (PriyangaRamasamy) [inline vpd-tool/src/vpd_tool.cpp:5]: why not tool_utils.hpp?

File: vpd-tool/src/vpd_tool.cpp [hunk 6]
```diff
@@ -8,13 +14,60 @@ int VpdTool::readKeyword(const std::string& i_fruPath,
...
+    {
+        types::DbusVariantType l_keywordValue;
+        if (i_onHardware)
+        {
+            // TODO: Implement read keyword's value from hardware
+        }
+        else
+        {
+            std::string l_inventoryObjectPath(constants::baseInventoryPath +
+                                              i_fruPath);
```
Comment (PriyangaRamasamy) [inline vpd-tool/src/vpd_tool.cpp:28]: what if the user provides eeprom path without giving -H option ? In that case Line 28 binds the given eeprom path with inventory prefix which will fail in next step.
please handle it

File: vpd-tool/src/vpd_tool_main.cpp [hunk 7]
```diff
@@ -47,38 +48,46 @@ int main(int argc, char** argv)
...
-    {
-        std::cerr << "Given EEPROM file path doesn't exist : " + l_vpdPath
-                  << std::endl;
-        return l_rc;
-    }
-
     (void)l_fileOption;
 
-    if (*l_readFlag)
+    if (l_readFlag->count() > 0)
```
Comment (PriyangaRamasamy) [inline vpd-tool/src/vpd_tool_main.cpp:75]: if user provides multiple -r in the command line, we should throw error.
so please check if the l_hardwareFlag->count() == 1 to proceed. Else throw unexpected command line option

File: include/utility/dbus_utility.hpp (hunk 3 above)
Comment (branupama) [reply include/utility/dbus_utility.hpp:109]: yes, instead of modifying in utils.hpp, I did in wrong file.
I removed this file changes.

File: vpd-tool/include/tool_constants.hpp [hunk 8]
```diff
@@ -9,5 +9,9 @@ namespace constants
 static constexpr auto KEYWORD_SIZE = 2;
 static constexpr auto RECORD_SIZE = 4;
 
+constexpr auto inventoryManagerService =
+    "xyz.openbmc_project.Inventory.Manager";
+constexpr auto baseInventoryPath = "/xyz/openbmc_project/inventory";
```
[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/include/tool_constants.hpp:15]: as this is path, I feel Prefix doesn't suits here.
variable name looks fine for me.

File: vpd-tool/src/vpd_tool.cpp (hunk 5 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool.cpp:5]: renames files which are duplicate names in this repo.
utils.hpp is only file in this repo.

File: vpd-tool/include/tool_constants.hpp [hunk 9]
```diff
@@ -9,5 +9,9 @@ namespace constants
...
+constexpr auto ipzVpdInf = "com.ibm.ipzvpd.";
```
[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/include/tool_constants.hpp:16]: done

File: vpd-tool/src/vpd_tool.cpp (hunk 6 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool.cpp:28]: if user providing eeprom path without choosing -H option, 
then this request is considered as dbus request.

user has to apply -H.

File: vpd-tool/src/vpd_tool.cpp (hunk 4 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool.cpp:39]: if the value is printable then value is printed in ASCII format,
if not value is printed in hex format.

File: vpd-tool/src/vpd_tool_main.cpp [hunk 10]
```diff
@@ -47,38 +48,46 @@ int main(int argc, char** argv)
 
     CLI11_PARSE(l_app, argc, argv);
 
-    if (*l_objectOption && l_vpdPath.empty())
+    if ((l_objectOption->count() > 0) && l_vpdPath.empty())
     {
         std::cout << "Given path is empty." << std::endl;
-        return -1;
+        return l_rc;
     }
```
[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp:56]: the below checks takes care of this.

File: vpd-tool/src/vpd_tool_main.cpp [hunk 11]
```diff
@@ -47,38 +48,46 @@ int main(int argc, char** argv)
...
-    if (*l_keywordOption &&
+    if ((l_keywordOption->count() > 0) &&
         (l_keywordName.size() != vpd::constants::KEYWORD_SIZE))
     {
         std::cerr << "Keyword " << l_keywordName << " is not supported."
                   << std::endl;
         return l_rc;
     }
 
-    if (*l_hardwareFlag && !std::filesystem::exists(l_vpdPath))
```
[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp:70]: yes.
hardware flag is checked for the applicable options only.

File: vpd-tool/src/vpd_tool_main.cpp (hunk 7 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp:75]: If the user mistakenly gives twice, this is still fine.

we are taking action as user chooses the option, instead of blocking the results or throwing an error.

File: vpd-tool/include/utils.hpp (hunk 1 above)
Comment (branupama) [reply vpd-tool/include/utils.hpp:21]: this API is duplicated from dbus_utility.
the statement looks fine for me.

**Review by souvik1914581 — APPROVED** [review]

File: vpd-tool/src/vpd_tool_main.cpp (hunk 7 above)
Comment (SunnySrivastava1984) [reply vpd-tool/src/vpd_tool_main.cpp:77]: To resolve this, should we check here how the standard command line operations work?

File: vpd-tool/src/vpd_tool.cpp (hunk 5 above)
Comment (SunnySrivastava1984) [reply vpd-tool/src/vpd_tool.cpp:5]: For consistency, I would agree with @PriyangaRamasamy

File: vpd-tool/src/vpd_tool.cpp [hunk 12]
```diff
@@ -1,20 +1,73 @@
...
+        {
+            std::string l_inventoryObjectPath(constants::baseInventoryPath +
+                                              i_vpdPath);
+
+            l_keywordValue = utils::readDbusProperty(
+                constants::inventoryManagerService, l_inventoryObjectPath,
+                constants::ipzVpdInfPrefix + i_recordName, i_keywordName);
+        }
+
+        if (const auto l_value =
```
Comment (SunnySrivastava1984) [inline vpd-tool/src/vpd_tool.cpp:35]: l_keywordValue can be empty as well, which shows that there was some error in Dbus operation? Should be try to identify that?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: vpd-tool/src/vpd_tool_main.cpp (hunk 7 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp:77]: yes, I checked few linux commands, it works with duplicate options given.

`anupama@gfwa601:~$ ls -lh -lh
total 1.5G
-rw-r--r--  1 anupama 1122842 154K Nov  7 10:21 50001001.json
-rwxr-xr-x  1 anupama 1122842  17K Nov 12 02:39 a.out
drwxr-xr-x  2 root    root    4.0K Jun 28 01:54 anupama_san5

anupama@gfwa601:~$ grep -nr -r "BUS" test/*
grep: test/factory.hpp.gch: binary file matches`

File: vpd-tool/src/vpd_tool.cpp (hunk 5 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool.cpp:5]: Updated

File: vpd-tool/src/vpd_tool.cpp (hunk 12 above)
Comment (branupama) [reply vpd-tool/src/vpd_tool.cpp:35]: updated.

File: vpd-tool/src/vpd_tool_main.cpp (hunk 7 above)
Comment (SunnySrivastava1984) [reply vpd-tool/src/vpd_tool_main.cpp:77]: In that case I am ok with your code.

File: vpd-tool/include/tool_constants.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply vpd-tool/include/tool_constants.hpp:15]: ok

File: vpd-tool/src/vpd_tool_main.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply vpd-tool/src/vpd_tool_main.cpp:70]: what i meant is, hardware flag is required for both read and write. so you can have it in common place.. not under if condition for read.

[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp:70]: H/w is used only under read and write currently.
and we only check H/w flag under read/write is chosen.

To keep it in common place we need to add a check again outside with which all flag uses h/w flag. with this remaining options has to go through which is unnecessary.

So keeping the flag locally.

File: vpd-tool/src/vpd_tool_main.cpp [hunk 13]
```diff
@@ -45,40 +46,50 @@ int main(int argc, char** argv)
...
-    }
-
     (void)l_fileOption;
 
-    if (*l_readFlag)
+    if (l_readFlag->count() > 0)
     {
-        // TODO: call read keyword implementation from here.
+        if ((l_hardwareFlag->count() > 0) &&
+            !std::filesystem::exists(l_vpdPath))
```
[the remark this answers is not in the document]
Comment (branupama) [reply vpd-tool/src/vpd_tool_main.cpp]: handled exception.

**Review by PriyangaRamasamy — APPROVED** [review]

**Review by RekhaAparna01 — APPROVED** [review]

File: vpd-tool/src/vpd_tool.cpp [hunk 14]
```diff
@@ -1,20 +1,76 @@
...
+                l_keywordValInJson.emplace(i_keywordName, l_keywordStrValue);
+                l_resultInJson.emplace(i_vpdPath, l_keywordValInJson);
+
+                utils::printJson(l_resultInJson);
+            }
+            else
+            {
+                // TODO: Write result to a given file path.
+            }
+            l_rc = 0;
```
Comment (SunnySrivastava1984) [inline vpd-tool/src/vpd_tool.cpp:57]: Just to send return code as success/failure from one common place this initialization has been done here?
At first sight it looks like a random magic number assignment. In case you want to have these kind of assignments, at least have a constant defined to depict its actual meaning. Here "0" does not mean Zero it means success.

Comment (SunnySrivastava1984) [reply vpd-tool/src/vpd_tool.cpp:57]: Add a TODO and create a defect for now, we will have to take it up at repo level.

File: vpd-tool/include/tool_utils.hpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+    try
+    {
+        auto l_bus = sdbusplus::bus::new_default();
+        auto l_method =
+            l_bus.new_method_call(i_serviceName.c_str(), i_objectPath.c_str(),
+                                  "org.freedesktop.DBus.Properties", "Get");
+        l_method.append(i_interface, i_property);
+
+        auto result = l_bus.call(l_method);
```
Comment (jinuthomas) [inline vpd-tool/include/tool_utils.hpp:61]: should you not use result.is_method_error() before reading the value ?

File: vpd-tool/src/vpd_tool_main.cpp [hunk 16]
```diff
@@ -45,40 +46,60 @@ int main(int argc, char** argv)
     auto l_hardwareFlag = l_app.add_flag("--Hardware, -H",
                                          "CAUTION: Developer only option.");
 
+    // ToDo: Take offset value from user for hardware path.
+
     CLI11_PARSE(l_app, argc, argv);
 
-    if (*l_objectOption && l_vpdPath.empty())
+    if ((l_objectOption->count() > 0) && l_vpdPath.empty())
```
Comment (jinuthomas) [inline vpd-tool/src/vpd_tool_main.cpp:53]: why this change ?

File: vpd-tool/src/vpd_tool_main.cpp [hunk 17]
```diff
@@ -45,40 +46,60 @@ int main(int argc, char** argv)
...
     {
         std::cout << "Given path is empty." << std::endl;
```
Comment (jinuthomas) [inline vpd-tool/src/vpd_tool_main.cpp:55]: given path or parameter is empty ?

File: vpd-tool/src/vpd_tool_main.cpp [hunk 18]
```diff
@@ -45,40 +46,60 @@ int main(int argc, char** argv)
...
-    if (*l_keywordOption &&
+    if ((l_keywordOption->count() > 0) &&
         (l_keywordName.size() != vpd::constants::KEYWORD_SIZE))
     {
         std::cerr << "Keyword " << l_keywordName << " is not supported."
                   << std::endl;
         return l_rc;
     }
 
-    if (*l_hardwareFlag && !std::filesystem::exists(l_vpdPath))
```
Comment (jinuthomas) [inline vpd-tool/src/vpd_tool_main.cpp:70]: could have used CLI::ExistingFile: Requires that the file exists if given.
if objectPath and hardware then Existing file , just FYI

**Review by jinuthomas — CHANGES_REQUESTED** [review]

**Review by jinuthomas — APPROVED** [review]
